system/jlib/jsocket.cpp

```diff
@@ -3313,82 +3417,57 @@ static bool decodeNumericIP(const char *text,unsigned *netaddr)
     return false;
 }
 
-static bool lookupHostAddress(const char *name,unsigned *netaddr)
+static void RecursionSafeLogErr(int ret, int ref, const char *msg, unsigned lineno, const char *name)
 {
-    // if IP4only or using MS V6 can only resolve IPv4 using 
-    static bool recursioncheck = false; // needed to stop error message recursing
-    unsigned retry=10;
-#if defined(__linux__) || defined (__APPLE__) || defined(getaddrinfo)
-    if (IP4only) {
-#else
+    static __thread bool recursioncheck = false; // needed to stop error message recursing
+    if (!recursioncheck)
     {
+        recursioncheck = true;
+        LogErr(ret, ref, msg, lineno, name);
+#ifdef _DEBUG
+        PrintStackReport();
 #endif
-        CriticalBlock c(hostnamesect);
-        hostent * entry = gethostbyname(name);
-        while (entry==NULL) {
-            if (retry--==0) {
-                if (!recursioncheck) {
-                    recursioncheck = true;
-                    LogErr(h_errno,1,"gethostbyname failed",__LINE__,name);
-                    recursioncheck = false;
-                }
-                return false;
-            }
-            {
-                CriticalUnblock ub(hostnamesect);
-                Sleep((10-retry)*100);
-            }
-            entry = gethostbyname(name);
-        }
-        if (entry->h_addr_list[0]) {
-            unsigned ptr = 0;
-            if (!PreferredSubnet.isNull()) {
-                for (;;) {
-                    ptr++;
-                    if (entry->h_addr_list[ptr]==NULL) {
-                        ptr = 0;
-                        break;
-                    }
-                    IpAddress ip;
-                    ip.setNetAddress(sizeof(unsigned),entry->h_addr_list[ptr]);
-                    if (PreferredSubnet.test(ip))
-                        break;
-                }
-            }
-            memcpy(&netaddr[3], entry->h_addr_list[ptr], sizeof(netaddr[3]));
-            netaddr[2] = 0xffff0000;
-            netaddr[1] = 0;
-            netaddr[0] = 0;
-            return true;
-        }
-        return false;
+        recursioncheck = false;
     }
-#if defined(__linux__) || defined (__APPLE__) || defined(getaddrinfo)
+}
+
+int getAddressInfo(const char *name, unsigned *netaddr, bool okToLogErr)
+{
     struct addrinfo hints;
-    memset(&hints,0,sizeof(hints));
-    struct addrinfo  *addrInfo = NULL;
-    for (;;) {
-        memset(&hints,0,sizeof(hints));
+    struct addrinfo *addrInfo = NULL;
+    int retry=10;
+    int retCode;
+
+    // each retry could take up to several seconds, depending on how DNS resolver is configured
+    // should a few specific non-zero return codes break out early from retry loop (EAI_NONAME) ?
+    while (true)
+    {
+        memset(&hints, 0, sizeof(hints));
```

Comment:
hint initialization and setting can be moved outside of the while() loop I think.